ocrd_keraslm/lib/rating.py

```diff
@@ -1,135 +1,605 @@
-from keras.utils import to_categorical
-from keras.models import Sequential, load_model
-from keras.layers import Dense
-from keras.layers import LSTM
-from keras.preprocessing.sequence import pad_sequences
-
-import click, numpy, pickle
+from keras.callbacks import Callback
+import click, numpy, pickle, codecs
+from random import shuffle
+from math import log, exp, ceil, floor
 
 class Rater(object):
+    '''A character-level RNN language model for rating text.
+    
+    Uses Keras to define, compile, train, run and test an RNN
+    (LSTM) language model on the (UTF-8) byte level. The model's
+    topology (layers depth, per-layer width, window length) can
+    be controlled before training.
 
-    def __init__(self):
-        '''
-        The constructor.
-        '''
+    To be used by stand-alone CLI (`scripts.train` for training,
+    `scripts.apply` for prediction, `scripts.test` for evaluation),
+    or OCR-D processing (`wrapper.ocrd_keraslm_rate`). 
 
+    Interfaces:
+    - `Rater.train`/`scripts.train` : file handles of byte sequences
+    - `Rater.test`/`scripts.test` : file handles of byte sequences
+    - `Rater.rate`/`scripts.apply` : character string
+    - `Rater.rate_once`/`wrapper.ocrd_keraslm_rate` : character string
+    - `Rater.rate_single`/`scripts.generate` or `wrapper.ocrd_keraslm_rate` : alternative list of bytes and states
+    '''
+    
+    def __init__(self):
         self.clear()
-
+    
     def clear(self):
-        '''
-        Resets rater.
-        '''
+        '''Reset model and set all parameters to their defaults.'''
+        
+        self.model = None
+        self.status = 0 # empty / compiled / trained?
+        
+        self.length = 0 # will be overwritten by CLI for train / by load model for rate/test
+        self.width = 0 # will be overwritten by CLI for train / by load model for rate/test
+        self.depth = 0 # will be overwritten by CLI for train / by load model for rate/test
+        self.length = 0 # will be overwritten by CLI for train / by load model for rate/test
+        
+        self.variable_length = False # also train on partially filled windows
+        self.stateful = True # keep states across batches within one text (implicit state transfer)
+        self.minibatch_size = 128 # will be overwritten by length if stateful
+        self.validation_split = 0.2 # fraction of training data to use for validation (generalization control)
+        
+        self.incremental = False # whether compiled with additional (initial) input state and (final) output state (explicit state transfer)
+    
+    def configure(self):
+        '''Define and compile model for the given parameters.'''
+        from keras.layers import Dense, TimeDistributed, Input
+        from keras.layers import LSTM, CuDNNLSTM
+        from keras import backend as K
+        from keras.models import Model
```

Comment:
Right, this could be seen as a matter of style. (If so, I don't mind changing.) But keeping them inside the function also might reduce memory requirement in other contexts. Those imports in `configure` make up nearly 120 MB resident. If the processor were embedded in a (multi-processing/threading) client-server architecture (as I think it will eventually), then wouldn't `KerasRate::process()` save those, because it never calls `configure` (as opposed to the constructor)? Admitted, I am not sure myself.